### PR TITLE
[SQL] Optimization to pull filters above joins

### LINKED ISSUE #6279: [SQL] Optimization to pull filters above joins across LOCAL views
Calcite has an optimization to pull filters above joins if possible.
We also do this when we synthesize joins.
But we don't have an optimization that pulls a filter defined after a view produced by a join - each view is optimized by Calcite separately.
It is not clear how frequently this would apply, so we could do some instrumentation to try to figure out.
This will also only work with local views.

### PR DESCRIPTION
**Pull request by mihaibudiu**
Fixes #6279 

There are calcite optimizations to pull predicates above joins if they only depend on one side of a join. But Calcite optimizes each view separately. This PR introduces an optimization which operates on the graph produced by stitching all views together. It does find a few predicates that can be pulled, most of them of the shape "IS NOT NULL".

### Describe Manual Test Plan

Ran all Java tests manually.

## Checklist

- [x] Unit tests added/updated

### CODE REVIEW

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 1]
```diff
@@ -11,6 +11,20 @@
...
+import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRawTuple;
+import org.dbsp.util.Logger;
+import org.dbsp.util.Maybe;
+import org.dbsp.util.Utilities;
+
+import java.util.List;
 
 /** Combine a {@link DBSPJoinOperator} followed by a {@link DBSPFilterOperator} into a
  * {@link DBSPJoinFilterMapOperator}.  Similar expansion for {@link DBSPLeftJoinOperator}. */
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java]: The class javadoc is now stale — this visitor no longer just combines `Join + Filter` into `JoinFilterMap`; it also pulls filters above the join. Please update the description to mention the new pull-above-join transformation and clarify which join kinds it applies to (only `DBSPJoinOperator` here — `DBSPLeftJoinOperator` and `DBSPStarJoinOperator` still fall through to the JoinFilterMap rewrite below). The `Similar expansion for DBSPLeftJoinOperator` sentence is misleading given the new code only covers inner joins.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 2]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+
+        @Override
+        public VisitDecision preorder(NoExpression expression) {
+            this.found = true;
+            return VisitDecision.STOP;
+        }
+
+        @Override
+        public VisitDecision preorder(DBSPType type) {
+            return VisitDecision.STOP;
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java:57]: Worth a short comment explaining *why* `preorder(DBSPType)` returns STOP: a `NoExpression` carries a `DBSPType` and its `accept` walks that type; without this short-circuit you'd descend into type structure that can't contain a `NoExpression` anyway. Future readers will wonder. Also: are you confident no `NoExpression` can be embedded inside a type (e.g. inside an array/map type's element type via some constant-folded expression that materialized in a type)? Worth a one-liner asserting the assumption.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 3]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+            cno.apply(expression);
+            return cno.found;
+        }
+    }
+
     @Override
     public void postorder(DBSPFilterOperator operator) {
         OutputPort source = this.mapped(operator.input());
         int inputFanout = this.getGraph().getFanout(operator.input().node());
         if (source.node().is(DBSPJoinOperator.class) && (inputFanout == 1)) {
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java:149]: This whole new block only handles `DBSPJoinOperator` (inner join). For `DBSPLeftJoinOperator` (handled in the `else if` further down) it is unsafe in general to pull a predicate referring to the right side above a LEFT JOIN — NULL-padded rows from non-matches would be filtered differently. Pulling a predicate that depends only on the left input above a LEFT JOIN is fine; pulling one that depends only on the right or only on the key is *not* equivalent. Could you add an explicit comment here noting that the LEFT/STAR cases deliberately skip this pull optimization, and (ideally) a regression test asserting that a LEFT JOIN + WHERE on the preserved side is *not* incorrectly transformed?

**Review by mythical-fred — COMMENTED** [review]
Nice optimization. Algorithm (substitute `NoExpression` and search for it post-inlining) is clever and looks sound for inner joins. A few things worth tightening before merge — see inline. Main asks: more tests (negative case, key-only push-to-both, multi-view), update the stale class javadoc, and split the unrelated logger-level change out (or call it out in the message).

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 4]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+            // Check if filter can be pushed to one input.  We use the following
+            // algorithm: we compute the function joinFunction \circ filterFunction
+            // and check whether it does not depend on one of the join inputs.
+            // For that purpose we apply it to a NoExpression on that input
+            // and check if the NoExpression appears in the resulting program.
+            Projection projection = new Projection(compiler, true, false);
+            DBSPClosureExpression joinFunction = join.getClosureFunction();
+            projection.apply(joinFunction);
+            DBSPClosureExpression filterFunction = operator.getClosureFunction();
+            if (projection.hasIoMap()) {
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java]: Why gate the whole pull-above-join transformation on `projection.hasIoMap()`? The optimization's correctness only requires that, after substituting `NoExpression` for one side, the resulting predicate be `NoExpression`-free. That property does not require the join function to be a pure projection — a non-projection join function (e.g. a cast on a key column) could still produce a side-independent predicate. Is the projection check just a conservative correctness condition because `applyAfter` + reduction is only reliable for projections, or is there a deeper reason? Worth a comment, and possibly worth relaxing later.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 5]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+
+                final DBSPVariablePath left = new DBSPTypeRawTuple(
+                        joinFunction.parameters[0].getType(),
+                        joinFunction.parameters[1].getType()).var();
+                final DBSPClosureExpression leftProjection = joinFunction.call(
+                        left.field(0),
+                        left.field(1),
+                        new NoExpression(joinFunction.parameters[2].getType()))
+                        .closure(left);
+                final DBSPClosureExpression leftFilterPredicate = filterFunction.applyAfter(compiler, leftProjection, Maybe.YES);
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java]: `Maybe.YES` here forces `applyAfter` to inline-and-reduce the composition (taking the `shouldInline = true` branch), which is what makes the `NoExpression` search meaningful — without reduction the `NoExpression` would always remain syntactically present inside a `let`. That's correct, but it's a subtle and load-bearing argument; please add a one-line comment saying *why* `YES` is needed (so a future cleanup doesn't switch it to `MAYBE`).

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 6]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+                if (!ContainsNoExpression.search(this.compiler, leftFilterPredicate)) {
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java]: The soundness of this whole technique rests on: "if `reduce` does not eliminate the `NoExpression`, then `NoExpression` survives in the result, AND if `reduce` *does* eliminate it, the predicate truly does not depend on that side." The second direction is the worrying one — could `reduce` constant-fold away a `NoExpression` *and* still produce a predicate that semantically depends on the absent value? E.g. an expression like `no AND false` reduces to `false`, but the original predicate `<rhs-expr> AND false` is also `false`, so that's actually safe. But what about `IS NULL(no)` — does the reducer have any short-circuit for `IS NULL` that could fold a `NoExpression` to a constant `true`/`false`? If so, this would silently miscompile. Could you add a comment about which kinds of reductions are known not to swallow `NoExpression` semantically, or convince yourself with a few adversarial unit cases?

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 7]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+                    this.addOperator(filter);
+                    rightInput = filter.outputPort();
+                    if (leftInput == null)
+                        leftInput = join.left();
+                }
+
+                // It is possible for a predicate to be pulled on both sides, if it only
+                // depends on the key
+                if (leftInput != null) {
+                    DBSPSimpleOperator newJoin = join.withInputs(List.of(leftInput, rightInput), true)
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java]: Minor: `withInputs(..., true)` — the `true` is the `force` flag on `DBSPSimpleOperator.withInputs`, which makes the operator be rebuilt even if the inputs and types look equal. That is right for the both-sides-pushed case where `leftInput` and `rightInput` may both still equal the original join inputs structurally, but pushing through a no-op filter — would never happen here (filter was added when `ContainsNoExpression` was false). Could you drop a quick comment about *why* `force=true` is needed, or whether it could be `false`?

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 8]
```diff
@@ -19,12 +33,110 @@ public FilterJoinVisitor(DBSPCompiler compiler, CircuitGraphs graphs) {
...
+                            .to(DBSPSimpleOperator.class);
+                    // No more filter after the join
+                    this.map(operator, newJoin);
+                    Logger.INSTANCE.belowLevel(this, 1)
+                                    .append("Pulled predicate before join ")
+                                    .appendSupplier(filterFunction::toString)
+                                    .newline();
+                    return;
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java]: Test gap: the only test (`issue6279`) covers exactly one case (predicate on the left side of an inner join). Given how much logic this PR introduces, the unit tests should also cover:
- predicate that depends on BOTH sides → must NOT be pulled (assert a `DBSPFilterOperator` *does* remain),
- predicate that depends only on the key column → pushed to BOTH sides (assert two filters appear, one above each input),
- predicate over a LEFT JOIN that touches the right (nullable) side → must NOT be pulled (regression for the LEFT-join soundness concern above),
- NULL-handling: e.g. `WHERE y IS NULL` and `WHERE y IS NOT NULL` on the inner-join case (you mentioned in the description that `IS NOT NULL` is the common shape being pulled),
- a non-projection join function (e.g. with a cast) so the `hasIoMap` early-out is exercised.

Mihai — your usual principle is "docs need backing tests" and the same applies here. One trivial happy path isn't enough scaffolding to prevent regressions of this optimization.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPFilterOperator.java [hunk 9]
```diff
@@ -42,6 +42,7 @@ public final class DBSPFilterOperator extends DBSPUnaryOperator implements ILine
     public DBSPFilterOperator(CalciteRelNode node, DBSPExpression condition, OutputPort input) {
         super(node, "filter", condition, input.outputType(), input.isMultiset(), input);
         this.checkResultType(condition, new DBSPTypeBool(CalciteEmptyRel.INSTANCE, false));
+        checkArgumentFunctionType(condition, input);
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPFilterOperator.java:45]: Defensive cleanup or latent bug fix? If you discovered that someone constructed a `DBSPFilterOperator` with a closure whose argument type didn't match the input row type while developing this PR (e.g. via the new `withInputs(..., true)` path), please mention it in the commit message — and ideally add a unit test that exercises the bad path so the assertion can't silently regress to never firing. If this is just hardening ("every other operator constructor does it, this one should too") then a one-liner in the commit message saying so is enough.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/CircuitCloneVisitor.java [hunk 10]
```diff
@@ -92,7 +92,7 @@ protected void map(OutputPort oldPort, OutputPort newPort, boolean add) {
                         () -> "Replacing operator with type\n" + oldPort.outputType() +
                                 " with new type\n" + newPort.outputType());
             }
-            Logger.INSTANCE.belowLevel(this, 1)
+            Logger.INSTANCE.belowLevel(this, 2)
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/CircuitCloneVisitor.java:95]: These two `belowLevel(this, 1) → belowLevel(this, 2)` changes are unrelated to the filter-above-join optimization. They're fine in isolation — `CircuitCloneVisitor` was apparently too chatty at level 1 — but they belong in a separate small commit (`Reduce CircuitCloneVisitor logging verbosity`) rather than hidden inside this PR. At minimum, mention them in the commit message.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegression2Tests.java [hunk 11]
```diff
@@ -1,5 +1,7 @@
 package org.dbsp.sqlCompiler.compiler.sql.simple;
 
+import org.dbsp.sqlCompiler.circuit.DBSPCircuit;
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegression2Tests.java]: Unused import — `DBSPCircuit` is not referenced anywhere in this file.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegression2Tests.java [hunk 12]
```diff
@@ -802,4 +804,35 @@ public void endVisit() {
...
+                  (5, 300);
+                """, """
+                 T.x | y | S.x | z
+                -----------------------
+                 1   | 5 | 1   | 100
+                 2   | 3 | 2   | 200""");
+        ccs.visit(new CircuitVisitor(ccs.compiler) {
+            @Override
+            public void postorder(DBSPFilterOperator node) {
+                Assert.fail("Shoudl have been eliminated");
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegression2Tests.java]: Typo: `Shoudl` → `Should`.

**Review by mythical-fred — APPROVED** [review]
Re-APPROVE on `320a2406d`.

The rewrite addresses essentially every point I raised on `70315bd5`:

- Class javadoc now mentions both pull-above-join and the existing JoinFilterMap fusion.
- `hasIoMap()` gate is gone entirely — the transformation is now driven purely by whether the substituted predicate still mentions `NoExpression`. Cleaner than what I asked for.
- `withInputs(..., true)` → `withInputs(..., false)`.
- Unused `DBSPCircuit` import and the `Shoudl` typo are fixed.
- Test gap is closed: `issue6279` now covers (a) left-side pull, (b) right-side pull, (c) both-sides key-only pull on inner join, (d) LEFT JOIN with key-only predicate pulled to both sides, and crucially (e) the LEFT JOIN + predicate-on-right negative case that asserts exactly one `DBSPLeftJoinFilterMapOperator` survives and nothing was incorrectly hoisted.

The LEFT JOIN regression in particular — the one I was most worried about silently miscompiling — is now nailed down by the post-order assertion. `Maybe.YES` is still load-bearing in `applyAfter(...)`; that's the only original concern not annotated with a comment, but the new test coverage proves the invariant it depends on. Good enough.

Nothing else to flag. Approved.

**Review by mythical-fred — APPROVED** [review]
Upgrade COMMENT → APPROVE on `320a2406d0`.

The force-push folds in the test expansion I asked for in my 2026-06-17 COMMENT review. `issue6279` now covers all five branches end-to-end with both result-table assertions and structural circuit-shape assertions (no leftover `DBSPFilterOperator`/`DBSPJoinFilterMapOperator`):

1. pull above left input (`WHERE y < 10`)
2. pull above right input (`WHERE z > 10`)
3. pull above both inputs via key-only predicate (`WHERE x < 10`)
4. key-only pull above LEFT JOIN — verifies `isLeft=true` still permits both-sides when the predicate is over the join key
5. the negative case — `WHERE z > 10` on `LEFT JOIN` correctly does NOT pull (leaves exactly one `DBSPLeftJoinFilterMapOperator`)

That negative case in particular is the one I cared about: it pins the asymmetry between INNER and LEFT JOIN at test level, so a future relaxation can't silently regress NULL-row semantics. The javadoc on `FilterJoinVisitor` is also rewritten into the bullet form matching the new behavior. `STAR`-side still falls through to `JoinFilterMap` (carryover from current code, not a regression of this PR).

Two small follow-up nits left non-blocking:
- The `Maybe.YES` argument to `applyAfter` is load-bearing here (it requests reduction) but undocumented at the call site. A one-line `// reduce composed expression` would help.
- `checkPredicatePull` is the third method in this codebase that uses the `NoExpression` substitution trick to detect operand-independence; if a fourth appears it's worth extracting to a `Predicates.dependsOn(side)` helper.

Neither is worth blocking on. Approving.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java [hunk 13]
```diff
@@ -11,20 +12,146 @@
...
+
+        final DBSPVariablePath left = new DBSPTypeRawTuple(
+                joinFunction.parameters[0].getType(),
+                joinFunction.parameters[1].getType()).var();
+        final DBSPClosureExpression leftProjection = joinFunction.call(
+                        left.field(0),
+                        left.field(1),
+                        new NoExpression(joinFunction.parameters[2].getType()))
+                .closure(left);
+        final DBSPClosureExpression leftFilterPredicate = filterFunction.applyAfter(compiler, leftProjection, Maybe.YES);
```
Comment (gz) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java:92]: what _is_ Maybe.YES

**Review by gz — APPROVED** [review]
what does it mean to "pull something above"

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java (hunk 13 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/FilterJoinVisitor.java:92]: Force inlining, otherwise it will look like `let a = f() in g()`, where the NoExpression will not disappear

**Comment by mihaibudiu** [comment]
This is not good enough, it loses comments, I will have to think of something.

**Comment by mihaibudiu** [comment]
> what does it mean to "pull something above"

A join may have a condition which can be evaluated directly on an input, like `T JOIN S on T.x = 2`. You can filter T before you join it, pulling the predicate from the join to the left input. This should always be better. Calcite does this, but cannot handle it when T is a VIEW computed by a different query - each query is optimized separately.
